invokeai/app/invocations/grounded_sam.py

```diff
@@ -0,0 +1,203 @@
+            grounding_dino_pipeline = pipeline(
+                model=str(model_path),
+                task="zero-shot-object-detection",
+                local_files_only=True,
+                # TODO(ryand): Setting the torch_dtype here doesn't work. Investigate whether fp16 is supported by the
+                # model, and figure out how to make it work in the pipeline.
+                # torch_dtype=TorchDevice.choose_torch_dtype(),
+            )
+            assert isinstance(grounding_dino_pipeline, ZeroShotObjectDetectionPipeline)
+            return GroundingDinoPipeline(grounding_dino_pipeline)
+
+        with context.models.load_remote_model(source=GROUNDING_DINO_MODEL_ID, loader=load_grounding_dino) as detector:
+            assert isinstance(detector, GroundingDinoPipeline)
+            return detector.detect(image=image, candidate_labels=labels, threshold=threshold)
+
+    def _segment(
+        self,
+        context: InvocationContext,
+        image: Image.Image,
+        detection_results: list[DetectionResult],
+    ) -> list[DetectionResult]:
+        """Use Segment Anything (SAM) to generate masks given an image + a set of bounding boxes."""
+
+        def load_sam_model(model_path: Path):
+            sam_model = AutoModelForMaskGeneration.from_pretrained(
+                model_path,
+                local_files_only=True,
+                # TODO(ryand): Setting the torch_dtype here doesn't work. Investigate whether fp16 is supported by the
+                # model, and figure out how to make it work in the pipeline.
+                # torch_dtype=TorchDevice.choose_torch_dtype(),
+            )
+            assert isinstance(sam_model, SamModel)
+
+            sam_processor = AutoProcessor.from_pretrained(model_path, local_files_only=True)
+            assert isinstance(sam_processor, SamProcessor)
+            return SegmentAnythingModel(sam_model=sam_model, sam_processor=sam_processor)
+
+        with (
+            context.models.load_remote_model(source=SEGMENT_ANYTHING_MODEL_ID, loader=load_sam_model) as sam_pipeline,
+        ):
+            assert isinstance(sam_pipeline, SegmentAnythingModel)
+
+            masks = sam_pipeline.segment(image=image, detection_results=detection_results)
+
+        masks = self._to_numpy_masks(masks)
+        masks = self._apply_polygon_refinement(masks)
+
+        for detection_result, mask in zip(detection_results, masks, strict=False):
+            detection_result.mask = mask
+
+        return detection_results
+
+    def _to_numpy_masks(self, masks: torch.Tensor) -> list[npt.NDArray[np.uint8]]:
+        """Convert the tensor output from the Segment Anything model to a list of numpy masks."""
+        masks = masks.cpu().float()
+        masks = masks.permute(0, 2, 3, 1)
+        masks = masks.mean(dim=-1)
+        masks = (masks > 0).int()
+        masks = masks.numpy().astype(np.uint8)
+        masks = list(masks)
+        return masks
```

Comment:
Ah yes... feels like this pattern still needs a better solution in the type checkers. When applying a series of transformations to a variable, I often find myself wanting to intentionally re-use the same variable name with different types to signal that memory can be freed immediately. I guess the current solution would be to explicitly `del` old variables as you go.

Not a big deal in this case though since there's just one major type change. I'll adopt the suggestion.

Reply:
Adopted the suggestion in [cec7399](https://github.com/invoke-ai/InvokeAI/pull/6701/commits/cec739936696013a74179f8e7f53963e1d89eedd), and improved further in [bcd1483](https://github.com/invoke-ai/InvokeAI/pull/6701/commits/bcd1483a14e2af4b5cede1c65318628388d46cb8)